File: Chapter 04/C#/Section_4_4_Anti_Patterns/After_Rich_Domain_Thin_Controller/After/After.Domain/Models/Order.cs
namespace After.Domain.Models
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Net;

    /// <summary>
    /// THE RICH DOMAIN MODEL
    /// ARCHITECTURE NOTE: This solves the "Anemic Domain" anti-pattern.
    /// In the "Before" state, the Controller calculated the total and
    /// applied discounts. Now, the Order class is responsible for its 
    /// own data integrity. 
    /// </summary>
    public class Order
    {
        private const decimal GOLD_DISCOUNT_RATE = 0.9m;
        // Encapsulation: External classes cannot arbitrarily change 
        // the Total or the Id. They must use the provided methods.
        public int Id { get; private set; }
        public decimal Total { get; private set; }
        public Customer Customer { get; private set; }

        // Encapsulation: Prevents external code from doing _items.Add() 
        // which would bypass our RecalculateTotal logic.
        private readonly List<Item> _items = new List<Item>();
        public IReadOnlyList<Item> Items => _items.AsReadOnly();

        // Expression-bodied members for calculated data 
        public bool IsEligibleForDiscount => Customer != null && Customer.Type == "Gold";

        // Sometimes architects keep a separate CustomerEmail on the Order for Historical Accuracy.
        // The Scenario: If a customer changes their email address next year, do you want their old orders to show the new email or the email used at the time of purchase?
        // The Decision: If you want the order to be a "snapshot" in time, you store the string separately. If you want it to always reflect the customer's current email, you  use the alias.
        // This is an example of there being no "right" or "wrong" answer, simply a business preference.  We will use the alias here.
        public string CustomerEmail => Customer.Email;

        // The Atomic Truth: Logic and data are now perfectly unified.
        public decimal TotalPrice => _items.Sum(i => i.Price * i.Quantity)
                                 * (IsEligibleForDiscount ? GOLD_DISCOUNT_RATE : 1);

        /// <summary>
        /// Initializes a new instance of the <see cref="Order"/> class using a Rich Domain approach.
        /// </summary>
        /// <param name="customer">The <see cref="Customer"/> placing the order. 
        /// Required to encapsulate business rules like discount eligibility.</param>
        /// <exception cref="ArgumentNullException">Thrown when the customer is null, 
        /// ensuring the model's data integrity from the moment of creation.</exception>
        /// <remarks>
        /// ARCHITECTURE NOTE: By injecting the full Customer entity instead of just an email string, 
        /// the Order gains the "context" needed to calculate its own TotalPrice. 
        /// This eliminates the need for the Application Layer to manually handle business rules.
        /// </remarks>
        public Order(Customer customer)
        {
            // One assignment sets up the entire relationship
            Customer = customer ?? throw new ArgumentNullException(nameof(customer));
            Id = new Random().Next(1000, 9999);
        }

        /// <summary>
        /// Behavior is now co-located with the data it mutates.
        /// </summary>
        public void AddItem(Item item, Customer customer)
        {
            // Business Rule: Prices must be positive
            if (item.Price <= 0)
            {
                throw new InvalidOperationException(
                    "Item price must be positive.");
            }
            _items.Add(item);
        }
    }
}

File: Chapter 04/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter04/section_4_4_anti_patterns/after_rich_domain_thin_controller/domain/models/Order.java
package com.grokkingsoftwarearchitecture.chapter04.section_4_4_anti_patterns.after_rich_domain_thin_controller.domain.models;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Random;

/**
 * THE RICH DOMAIN
 * ARCHITECTURE NOTE: This solves the "Anemic Domain" anti-pattern.
 * In the "Before" state, the Controller calculated the total and
 * applied discounts. Now, the Order class is responsible for its 
 * own data integrity. 
 */
public class Order {
    private static final double GOLD_DISCOUNT_RATE = 0.9;
    
    // Encapsulation: External classes cannot arbitrarily change 
    // the total or the id. They must use the provided methods.
    private int id;
    private Customer customer;
    
    // Encapsulation: Prevents external code from doing items.add() 
    // which would bypass our recalculateTotal logic.
    private final List<Item> items = new ArrayList<>();

    /**
     * ARCHITECTURE NOTE: By injecting the full Customer entity instead of just 
     * an email string, the Order gains the "context" needed to calculate 
     * its own TotalPrice.
     */
    public Order(Customer customer) {
        if (customer == null) throw new IllegalArgumentException("Customer cannot be null");
        this.customer = customer;
        this.id = new Random().nextInt(9000) + 1000;
    }

    public int getId() { return id; }
    public Customer getCustomer() { return customer; }

    /** Encapsulation: Prevents external code from bypassing our business rules. */
    public List<Item> getItems() { return Collections.unmodifiableList(items); }

    public boolean isEligibleForDiscount() {
        return customer != null && "Gold".equals(customer.getType());
    }

    /**
     * ARCHITECTURE NOTE: We use an alias here to reflect the customer's current email.
     * If the business required a 'snapshot', we would store this as a separate string.
     */
    public String getCustomerEmail() { return customer.getEmail(); }

    /**
     * THE ATOMIC TRUTH: Logic and data are now perfectly unified.
     * This replaces the manual 'RecalculateTotal' method.
     */
    public double getTotalPrice() {
        double sum = items.stream().mapToDouble(i -> i.getPrice() * i.getQuantity()).sum();
        return isEligibleForDiscount() ? sum * GOLD_DISCOUNT_RATE : sum;
    }

    /**
     * Behavior is now co-located with the data it mutates.
     */
    public void addItem(Item item) {
        // Business Rule: Prices must be positive
        if (item.getPrice() <= 0) {
            throw new IllegalStateException(
                "Item price must be positive.");
        }   
        items.add(item);
    }
}